src/OpenApi/src/Extensions/JsonObjectSchemaExtensions.cs

```diff
@@ -0,0 +1,279 @@
+    {
+        foreach (var attribute in validationAttributes)
+        {
+            if (attribute is Base64StringAttribute)
+            {
+                schema["format"] = "byte";
+            }
+            if (attribute is RangeAttribute rangeAttribute)
+            {
+                schema["minimum"] = decimal.Parse(rangeAttribute.Minimum.ToString()!, CultureInfo.InvariantCulture);
+                schema["maximum"] = decimal.Parse(rangeAttribute.Maximum.ToString()!, CultureInfo.InvariantCulture);
+            }
+            if (attribute is RegularExpressionAttribute regularExpressionAttribute)
+            {
+                schema["pattern"] = regularExpressionAttribute.Pattern;
+            }
+            if (attribute is MaxLengthAttribute maxLengthAttribute)
+            {
+                schema["maxLength"] = maxLengthAttribute.Length;
+            }
+            if (attribute is MinLengthAttribute minLengthAttribute)
+            {
+                schema["minLength"] = minLengthAttribute.Length;
+            }
+            if (attribute is StringLengthAttribute stringLengthAttribute)
+            {
+                schema["minimum"] = stringLengthAttribute.MinimumLength;
+                schema["maximum"] = stringLengthAttribute.MaximumLength;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Applies the primitive types and formats to the schema based on the type.
+    /// </summary>
+    /// <remarks>
+    /// OpenAPI v3 requires support for the format keyword in generated types. Because the
+    /// underlying schema generator does not support this, we need to manually apply the
+    /// supported formats to the schemas associated with the generated type.
+    ///
+    /// Whereas JsonSchema represents nullable types via `type: ["string", "null"]`, OpenAPI
+    /// v3 exposes a nullable property on the schema. This method will set the nullable property
+    /// based on whether the underlying schema generator returned an array type containing "null" to
+    /// represent a nullable type or if the type was denoted as nullable from our lookup cache.
+    ///
+    /// Note that this method targets <see cref="JsonObject"/> and not <see cref="OpenApiSchema"/> because
+    /// it is is designed to be invoked via the `OnGenerated` callback in the underlying schema generator as
+    /// opposed to after the generated schemas have been mapped to OpenAPI schemas.
+    /// </remarks>
+    /// <param name="schema">The <see cref="JsonObject"/> produced by the underlying schema generator.</param>
+    /// <param name="type">The <see cref="Type"/> associated with the <see paramref="schema"/>.</param>
+    internal static void ApplyPrimitiveTypesAndFormats(this JsonObject schema, Type type)
+    {
+        if (_simpleTypeToOpenApiSchema.TryGetValue(type, out var openApiSchema))
+        {
+            schema["nullable"] = openApiSchema.Nullable || (schema["type"] is JsonArray schemaType
```

Comment:
Aren't we setting `schema["type"]` on the next line?  Is this a mutation, rather than an initialization?

Reply:
Yes, it's a mutation of the value generated by STJ's schema generator rather than an initialization. The remarks in the function comment clarify this but because OpenAPI schema doesn't support the _exact_ semantics for `type` as JSON schema we end up having to map things.

JSON schema knows:

```
type: ["int", "null"]
```

Whereas OpenAPI v3 knows:

```
type: "int",
nullable: true
```